modular_skyrat/modules/vore/code/datums/vore_prefs.dm

```diff
@@ -0,0 +1,584 @@
+
+	var/char_vars
+
+	var/needs_update = UPDATE_ALL|UPDATE_STATIC_DATA
+
+/datum/vore_prefs/New(client/holder = null, datum/preferences/_prefs=null)
+	if (!holder || !_prefs)
+		return
+	prefs = _prefs
+	path = "data/player_saves/[holder.ckey[1]]/[holder.ckey]/vore.sav"
+	if (fexists(path))
+		load_slotted_prefs(TRUE)
+	else
+		save_prefs()
+
+/datum/vore_prefs/proc/load_prefs(slot = null, read_selected=FALSE)
+	if (isnull(slot))
+		slot = prefs?.default_slot || 0
+	if (!path || !fexists(path))
+		return FALSE
+	var/savefile/S = new /savefile(path)
+	if (!S)
+		return FALSE
+	. = TRUE
+	S.cd = "/"
+	READ_FILE(S["vore_enabled"], vore_enabled)
+	//var/version
+	//READ_FILE(S["savefile_version"], version)
+	//do any sanity checks for version here in the future if you need to (create a new proc for it)
+	S.cd = "/char[slot]"
+	READ_FILE(S["vore_toggles"], vore_toggles)
+	READ_FILE(S["char_vars"], char_vars)
+	READ_FILE(S["bellies"], bellies)
+	if (read_selected)
+		READ_FILE(S["selected_belly"], selected_belly)
+	has_unsaved = FALSE
+	needs_update |= UPDATE_ALL
+	sanitize_vars()
+
+/datum/vore_prefs/proc/save_prefs(slot = null)
+	if (isnull(slot))
+		slot = prefs?.default_slot || 0
+	if (!path)
+		return FALSE
+	var/savefile/S = new /savefile(path)
+	if (!S)
+		return FALSE
+	. = TRUE
+	S.cd = "/"
+	WRITE_FILE(S["vore_enabled"], vore_enabled)
+	WRITE_FILE(S["savefile_version"], VORE_SAVEFILE_VERSION)
+	S.cd = "/char[slot]"
+	sanitize_vars()
+	WRITE_FILE(S["vore_toggles"], vore_toggles)
+	WRITE_FILE(S["selected_belly"], selected_belly)
+	WRITE_FILE(S["char_vars"], char_vars)
+	WRITE_FILE(S["bellies"], bellies)
+	has_unsaved = FALSE
+
+/datum/vore_prefs/proc/sanitize_vars()
+	//statics
+	var/static/list/default_vore_toggles = list(VORE_MECHANICS_TOGGLES = VORE_MECHANICS_TOGGLES_DEFAULT,\
+												VORE_CHAT_TOGGLES = VORE_CHAT_TOGGLES_DEFAULT/*,\
+												VORE_SOUND_TOGGLES = VORE_SOUND_TOGGLES_DEFAULT*/)
+	var/static/list/character_vars = list("tastes_of" = "nothing in particular")
+	var/static/list/default_belly_vars = default_belly_info()
+
+	if (isnull(vore_toggles) || !islist(vore_toggles))
+		vore_toggles = default_vore_toggles.Copy()
+	else
+		for (var/item in default_vore_toggles)
+			if (!(item in vore_toggles))
+				vore_toggles[item] = default_vore_toggles[item]
+	if (isnull(char_vars) || !islist(char_vars))
+		char_vars = character_vars.Copy()
+	else
+		for (var/item in character_vars)
+			if (!(item in char_vars))
+				char_vars[item] = character_vars[item]
+	if (!bellies || !bellies.len)
+		bellies = list()
+		bellies.Add(list(default_belly_info()))
+	else
+		for (var/belly in bellies)
+			for (var/item in default_belly_vars)
+				if (!(item in belly))
+					belly[item] = default_belly_vars[item]
+	selected_belly = clamp(selected_belly, 1, bellies.len)
+
+/datum/vore_prefs/proc/load_slotted_prefs(read_selected=FALSE, slot_from_prefs=null)
+	var/datum/component/vore/vore = prefs?.parent?.mob?.GetComponent(/datum/component/vore)
```

Comment:
get component bad, use signals, etc.

Reply:
Don't use signals like procs. Get component is fine, in rare cases, which this is one.

Reply:
AFAIK signals can only return bitflags, so no. 👎